### DISCUSSION TITLE
MonkeyPoll Ideas

### DISCUSSION THREAD

**Discussion by pathumd** [yehandesilva/monkeyPoll#4] [Ideas]
After thinking about this project for some time, here are some general thoughts I have on how the project could work:

I think that there should be three entities: A survey, a question (where there are three different types of questions: A text-based, numerical range, and selection), and account (for survey creators). In my opinion, I think the user should not be required to create an account to fill in a survey. Instead, the survey creator should create an account, so that they can log in, create surveys, close existing surveys, and see analytical results from closed surveys. Once a survey is created, the creator can share its unique link (the e.g. surveys/id?=434543) to which anyone can access and fill out.

In terms of verifying that the same user doesn't submit multiple submissions, we could start off by ensuring no submission with the user's first name and last name exists as a response for that particular survey. And in terms of closing surveys, a persisted attribute of the survey entity could be a boolean, indicating whether the form is open or closed. When the survey creator closes the survey, this flag will be updated. When the user accesses the form link to submit a response, the controller responsible for returning the form's view can handle checking whether the retrieved form is open or closed. If closed, the user should be shown an error message saying e.g. "The requested survey is no longer open".

**Comment by yehandesilva** [comment]
We have 2 requirements here:
1) Users can fill out surveys without signing-in.
2) Prevent users from making multiple responses to the same form.

Using first and last names as unique identifiers will not work because different people can share the same common name. If we want to ensure both requirements, we would need to somehow store user info with their response to ensure they can't respond again. But how can we do that if they are not going to sign in? We would have to store their IP in that case to prevent responses from the same IP address. 

I would say its better to stick with only one of the requirements.

**Reply by pathumd** [reply]
Agreed, using first and last names won't work as unique identifiers. However, forcing the user to specify their email address when submitting a survey may be a good way to uniquely identify a user.

**Reply by yehandesilva** [reply]
Yes, specifying email address would fulfuill both requirements and is better than storing IP addresses.
